CARDS-1337: Enable patient authentication

sdumitriu opened the pull request:
What's included:
- a service for generating tokens
- an authenticator that accepts tokens for login
- a `patient` user that is used for all patients
- access rights that **only** allow a visiting patient read access to the Visit and Patient subjects, and read-write access to **existing** forms for the visit
- a temporary helper script that can be used for manually generating tokens; this should be removed after testing

How it works:

- start with `./start_cards.sh --project cards4proms --dev --permissions trusted`
- as admin, create two Patient subjects (alice, bob), each with two Visit sub-subjects (a1, a2, b1, b2)
- create a Visit Information form for a1, leave the visit date empty
- create a Visit Information form for a2, set a visit date in the future
- create a Visit Information form for b1, set a visit date in the past
- do **not** create a Visit Information form for b2
- create other forms for some of these visits
- to create a token for a visit, access the Visit subject, and in the URL remove `/content.html` and append `.token.html` at the end, e.g. `http://localhost:8080/content.html/Subjects/0542e731-575d-4e06-bc83-f79f7f300166/7a96f652-cff6-4701-9b27-55a2c2409034` becomes `http://localhost:8080/Subjects/0542e731-575d-4e06-bc83-f79f7f300166/7a96f652-cff6-4701-9b27-55a2c2409034.token.html`
- the token will be printed in the page
- tokens are valid for: a1 and b2 10 minutes from creation, a2 until the visit date specified in the form, b1 should already be expired and not work at all
- to log in using a token, append `?auth_token=the token` to the URL, e.g. `http://localhost:8080/?auth_token=dbbc84ba-aaf0-4d75-8581-27ff194d7c99_08fa82bc722d100424db2f52c3b90c374e8cdb33c7007bc0`
- once logged in with a token, you should:
    - appear logged in, a `P` avatar indicating you are `patient`
    - have access to the homepage, but no data is shown, `Error - Access to data is pending the approval of your account` is displayed instead
    - have access to the correct Patient and Visit subjects (by manually copy-pasting the URL), but not to other visits or other patients
        - on the patient/visit pages, it's OK to have `Error obtaining subject data: 404 Not Found`
    - have access to the correct forms (by copy-pasting the URLs), can view, edit, save, re-edit
    - I didn't test, since there are no such questions in proms mode, but do vocabularies work?
    - after ten minutes, trying to access using the expired token should fail, re-displaying the login page
- using the proms UI (`/Proms.html/Cardio`) does **not** work at the moment

Things to do:

1. [x] If a token is expired, right now we just show the login page; is it worth spending time to instead display a message about the expired link? - [CARDS-1548](https://phenotips.atlassian.net/browse/CARDS-1548)
2. Logging in with an expired token will delete it, but there's no way to periodically delete expired tokens. - [CARDS-1549](https://phenotips.atlassian.net/browse/CARDS-1549)
3. [x] Looks like logging out doesn't work, **needs to be fixed in this PR**
4. We should prevent access to the regular homepage if a patient is logged in, they should only have access to the Proms UI once it's actually working - [CARDS-1339](https://phenotips.atlassian.net/browse/CARDS-1339)
5. When sending emails to the patient, call `tokenManager.create(date of the visit + 2hours, Collections.singletonMap("cards:visit", "uuid of the visit node")).getToken()` to get the login token, and append it to the URL. - to be considered in [CARDS-1422](https://phenotips.atlassian.net/browse/CARDS-1422) + [CARDS-1424](https://phenotips.atlassian.net/browse/CARDS-1424).
6. The proms UI needs to be reworked to actually get the data from the server instead of trying to create new forms for the fake test patients. - the backend support for this is in [CARDS-1349](https://phenotips.atlassian.net/browse/CARDS-1349) + [CARDS-1484](https://phenotips.atlassian.net/browse/CARDS-1484); UI will be adjusted in [CARDS-1550](https://phenotips.atlassian.net/browse/CARDS-1550).
7. [x] Disallow changing the subject (deny write access to form/subject)

sdumitriu committed:
CARDS-1337: Enable patient authentication
Add a TokenManager service, along with a custom implementation of Oak's TokenInfo

sdumitriu committed:
CARDS-1337: Enable patient authentication
Use our own token implementation in Oak's token-based authentication module

sdumitriu committed:
CARDS-1337: Enable patient authentication
Add Sling support for logging in using CARDS tokens

sdumitriu committed:
CARDS-1337: Enable patient authentication
Implement custom ACL restriction that only allows access to a specific visit

sdumitriu committed:
CARDS-1337: Enable patient authentication
Add the new authentication module to the PROMS project
Configure rights for a patient user

sdumitriu committed:
CARDS-1337: Enable patient authentication
Temporary helper script for creating a token

sashaandjic commented:
I cannot sign out when I'm logged as patient.

marta- commented:
> I cannot sign out when I'm logged as patient. Should change password option be available for patients? ![image](https://user-images.githubusercontent.com/23243244/145893902-2e53100a-e30c-400d-9898-e2d1bdb36177.png)

Patients will not have access to this UI, this will be ensured by properly set permissions in https://phenotips.atlassian.net/browse/CARDS-1339 .

sashaandjic commented:
After login with `?auth_token=the token`, should I be able to create a new questionnaire for the given subject?
It allows me to select the questionnaire and the subject for which I created token, but then I get the login dialog.
![image](https://user-images.githubusercontent.com/23243244/146251644-fedd3841-968c-4d0b-afbf-eaa26c56a54d.png)

sashaandjic commented:
When I click on statistics it gives me the data related to the subject to whom I created a token. However, when I click on the graph I get this:
![image](https://user-images.githubusercontent.com/23243244/146252839-d67f8157-d29f-4163-abd2-8030e49f573a.png)

sashaandjic commented:
I created a token for patient who has two visits. Each visit has data that should contribute to the Stats. When I click on `Statistics` I get 2 blank graphs. I am assuming this is correct, as the token is not applicable to the visits.
![image](https://user-images.githubusercontent.com/23243244/146254142-8b7d1511-b4b1-41dc-915a-aee296157693.png)

sashaandjic commented:
> * I didn't test, since there are no such questions in proms mode, but do vocabularies work?
Yes. Tested.

sdumitriu commented:
> After login with ?auth_token=the token, should I be able to create a new questionnaire for the given subject?
It allows me to clik on the (+) button, select questionnaire and only the subject for which I created token, but then I get the login dialog.

No, patients can only edit existing forms. They cannot create any new subjects/forms, and they should not be able to delete an existing form that they have access to. One thing that is allowed but shouldn't be is changing the subject. I'll fix that.

> Also, the login dialog appears after attempting to delete the subject.

The delete button should not be present. This is kind of a known bug, our UI doesn't take into account the actual rights that the user has, and always displays edit/delete buttons, even if the user cannot actually do that.

But this is irrelevant, the patients will not be able to access this UI at all once we properly implement the Proms UI and restrict access to the dashboard.

> When I click on Statistics it gives me the data related to the subject to whom I created a token.
> However, when I click on the graph I get this (white screen):

Patients should not have access to statistics at all, but it's not a big deal if they can fetch the stats data, if that contains only the accessible forms (the ones for the current visit).

And the statistics UI will not be accessible to patients.

> I created a token for patient who has two visits. Each visit has data that should contribute to the Stats. When I click on Statistics I get 2 blank graphs. I am assuming this is correct, as the token is not applicable to the visits.

Correct that the stats should not be shown.

But tokens should not be created for anything other than a visit. I did not prevent this in the code, though, since we may want to add other kinds of token usecases in the future, and the manual token generation will not be available, all tokens will be created automatically when sending emails to the patient.

> Question: Dose token created for patient ever expire?

In ten minutes.

sdumitriu committed:
CARDS-1337: Enable patient authentication
Disallow changing the subject or the questionnaire of an accessible form

sdumitriu committed:
CARDS-1337: Enable patient authentication
Fix logout not working

sdumitriu committed:
CARDS-1548: Support for expired links
- added a "link expired" page
CARDS-1337: Enable patient authentication
- moved the token authentication module into the core modules and enable it by default

sashaandjic commented:
https://user-images.githubusercontent.com/23243244/147788587-9c80071b-41ff-48a7-af45-59ee86645f60.mp4

sdumitriu committed:
CARDS-1337: Enable patient authentication
Enable the token authentication module

sashaandjic commented:
Not sure if the grammar is correct here. Perhaps "The link has expired." or "The token has expired.".
<img width="880" alt="image" src="https://user-images.githubusercontent.com/23243244/147834399-14fbb8b2-61a0-4f99-8ba3-c0374669c6a0.png">

sashaandjic commented:
- use token to login
-  open Visit form and 
- click on `Change subject`
- click on the NEW button in the Change subject dialog
- login dialog pops up

<img width="142" alt="image" src="https://user-images.githubusercontent.com/23243244/147834484-cdf82508-99ea-4062-8bd0-aed695beeb0a.png">

sashaandjic commented:
With the `visit a1` token login I am unable to access `visit a1`
<img width="945" alt="image" src="https://user-images.githubusercontent.com/23243244/147834895-e63e83b2-738b-43ad-bc4e-d41230399dd2.png">

 however I can access all `visit a1's` forms.
<img width="945" alt="image" src="https://user-images.githubusercontent.com/23243244/147834853-569de082-0a1e-46a1-949d-3927e3176a87.png">

Attempting to access any other form gives me 404 err. However, attempting to access any other visit seem to get me stuck.
<img width="949" alt="image" src="https://user-images.githubusercontent.com/23243244/147834880-a6b9bec1-7e14-42d7-bb41-e66737ae37fb.png">

sashaandjic commented:
Stats are available just for the visit for which the token was created. When I click on stat's graph, I get this:
<img width="759" alt="image" src="https://user-images.githubusercontent.com/23243244/147835149-c689b79b-bb96-408c-8e74-fc5d7d92a1e7.png">

sdumitriu commented:
> Suggestion: Disable `Change subject` and `Delete form` options.

The whole UI will be inaccessible to patients, only the PROMS UI will be available to them, and that one is not yet working.

But that may also affects regular users depending on their rights, so it will have to be fixed at some point, but not in this PR.

> With the visit a1 token login I am unable to access visit a1 page ( this was expected afair)
> however I can access, view and edit all visit a1's forms.
> Attempting to access any other form gives me correctly an 404 err.

Yep, as expected (rightswise, and the UI errors should be ignored).

> However, attempting to access any other visit seem to get me stuck.

Another "this is not the UI for patients" symptom, not a problem for now.

> Stats are available just for the visit for which the token was created. When I click on stat's graph, I get this:

Expected, patients are not allowed to perform queries.

marta- commented on modules/token-authentication/src/main/java/io/uhndata/cards/auth/tokenguest/impl/PatientVisitRestrictionPattern.java:
```diff
@@ -0,0 +1,137 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.uhndata.cards.auth.tokenguest.impl;
```
Minor comment: this restriction is pretty specific to the DATA Pro workflow, so it should probably be moved out of the generic `auth.tokenguest` package.

sdumitriu committed:
CARDS-1337: Enable patient authentication
Move token generation helper script in the correct module

sdumitriu committed:
CARDS-1337: Enable patient authentication
Renamed java package from `tokenguest` to `token`

sdumitriu committed:
CARDS-1337: Enable patient authentication
Make the visit ACL restriction more generic, it should not mention ePRO-specific visits, it should instead apply to any kind of subject.
Also move it in the cards-permissions module since it is generic enough.

sdumitriu committed:
CARDS-1337: Enable patient authentication
Improve the "link expired" message

sashaandjic commented:
<img width="917" alt="image" src="https://user-images.githubusercontent.com/23243244/148435987-5b99d67a-d896-40c3-9283-65353ff15dcf.png">

marta- reviewed (approved):
Code looks much better after refactoring. Didn't re-test.

sashaandjic reviewed (approved):
re-tested
all good!
